Initial local echo support

pixlwave opened the pull request:
Shows local echoes for messages, reactions and images by adding `outgoingMessages` from the SDK to the end of `events()`. Works well for free (I initially thought I'd have to have an echo store somewhere and keep it in sync).

Currently images just show the placeholder (which I've extended to include a solid colour when there's no blurhash) as they don't get a URL until they sync back - could potentially store the jpeg data somewhere and use that, but wasn't sure of the best place to do that for now.

Also messages do a little dance when replacing as I think the bubble view is displayed then then `sizeToFit` gets called on the label.

One thing to figure out before making this good to go is that if a message send fails, it just sits at the bottom (even after a force quit). Will leave edits and redactions for another PR.

pixlwave reviewed (commented):
Ok never mind that comment, this extension isn't needed, now that I've discovered that you *can* use a viewbuilder as a placeholder 👍

kiliankoe commented:
This is working nicely, great job! For a future optimization it might be cool to not save the outgoing events in the same event collection, but separately, then we can render them differently as well (maybe a lighter background color or small spinner attached or something along those lines). But this looks great for now 👍

kiliankoe commented on Nio/Conversations/Event Views/MessageView/MediaEventView.swift:
```diff
@@ -42,22 +42,24 @@ struct MediaEventView: View {
+        if let size = model.size,
+           let blurhash = model.blurhash,
+           let img = UIImage(blurHash: blurhash, size: size) {
+            Image(uiImage: img)
```
It would probably look nice to put the same mask as for the main image view (`.mask(RoundedRectangle(cornerRadius: 15))`) on the placeholder as well.

pixlwave replied:
Fixed! The placeholder inherits that mask, however I'd set the aspect ratio on the placeholder and not the image view so it didn't fill to the corners. Looks good now 😎

kiliankoe replied:
Oh nice, that explains it. Looks good!

pixlwave commented:
I did originally look at keeping `eventCache` and `outgoingEvents` separate, however when testing with matrix.org sometimes the delay between the message being successfully delivered (the point at which the callback is run) and the message being received back from the server was pretty large.

Turns out `MXEvent` has a `sentState` property which includes sending/sent/failed amongst other things, so I've mocked that up in the last commit. This way we can have a retry/cancel action on a send failure.

kiliankoe commented:
Ah fantastic, that's even better!

kiliankoe reviewed (approved):
Everything looks good to me 👍

kiliankoe commented:
This is such a fantastic little thing to have in the app, I totally underestimated how much better it would feel to have messages not disappear for even a small moment, but now everything feels so snappy!

pixlwave commented:
Sorry, I just added one more commit after you approved 😕.

I was bothered that if a message failed to send it would hang around at the bottom with no way to remove it. Now the context menu removes it from the outgoing messages.

I've also added the context menu to `MediaEventView` because that was the only place I ever saw it fail (and the most annoying because you have this big grey box blocking half your window). One tiny annoyance is that reactions haven't been implemented on that yet. I can set `canReact` for media to false for now with a Todo or just ignore it?

kiliankoe commented on Nio/Conversations/RoomView.swift:
```diff
@@ -121,7 +121,13 @@ struct RoomView: View {
                                     onEdit: { self.edit(event: event) },
-                                    onRedact: { self.eventToRedact = event.eventId }))
+                                    onRedact: {
+                                        if event.sentState == MXEventSentStateFailed {
```
Would it maybe make sense to not check if it's failed, but to check if it's anything but successfully sent here?

kiliankoe commented:
Eh, we can fix reactions for media events later 😅 

Thanks for the additional change, it definitely makes sense to support that 👍

pixlwave commented:
I was just trying to find how to reply to that last review you made on which sent states to remove. My only logic for restricting it to failed was that if I turned off my wifi, sent a message and then removed that from the outgoing messages, it still got sent when I connected again 🤷‍♂️. Not sure I understand why, but I guess that's something else to look into at some point.

kiliankoe commented:
I deleted the comment right after submitting it 😅 I thought I got confused there, since the code should be handling the other eventualities as well as far as I can tell. When cutting the internet connection (the [network link conditioner](https://nshipster.com/network-link-conditioner/) is a great option here btw), I tried removing a message in transit and it didn't disappear, but apparently got a sent state (it was no longer at half opacity). I never received it on the other side though after re-enabling the internet connection. Not sure though where the issue was, I wanted to play around with that a bit more.

pixlwave commented:
Oh, I'll take a look into the link conditioner 😄
